ip: Support static IP address on with auto ip stack

cathay4t opened the pull request:
Introducing two properties for IP address:
 * `valid-left`: seconds for IP address been valid. `forever` means
   static.
 * `preferred_left`: seconds for IP address been preferred. `forever`
   means static.

With above properties, we can identify which certain IP address is
retrieved from DHCP/Autoconf so we don't convert dynamic IP address to
static when user apply the queried state back.

With this patch, user is allowed to set static IP address with
`dhcp: true` and/or `autoconf: true`.

When desire state has DHCP/Autoconf disabled but all desired IP
addresses are dynamic, nmstate will convert those dynamic IP to static.
This is common use case where user just copy retrieved dynamic IP and
hoping nmstate can convert it to static.

Integration test cases included.

cgwalters commented on rust/src/lib/ip.rs:
```diff
@@ -274,18 +271,42 @@ impl InterfaceIpv4 {
+                    }
+                }
+            }
+        } else if let Some(true) = self.addresses.as_ref().map(Vec::is_empty) {
```
```suggestion
        } else if self.addresses.as_ref().map_or(false, Vec::is_empty) {
```

cathay4t replied:
Thanks for the suggestions. Applied.

cgwalters commented on rust/src/lib/ip.rs:
```diff
@@ -274,18 +271,42 @@ impl InterfaceIpv4 {
+            }
+        } else if let Some(true) = self.addresses.as_ref().map(Vec::is_empty) {
+            // Empty address should equal to disabled IPv4 stack
+            if self.enabled {
                 if is_desired {
```
```suggestion
            if self.enabled && is_desired
```

cgwalters commented on rust/src/lib/ip.rs:
```diff
@@ -744,10 +797,36 @@ impl Default for InterfaceIpAddr {
+                "{}/{} valid_left {} preferred_left {}",
+                self.ip,
+                self.prefix_length,
+                self.valid_left.as_deref().unwrap_or("forever"),
```
I think associated consts are good for this type of stuff, see below

cgwalters commented on rust/src/lib/ip.rs:
```diff
@@ -744,10 +797,36 @@ impl Default for InterfaceIpAddr {
+}
+
+impl InterfaceIpAddr {
+    pub(crate) fn is_auto(&self) -> bool {
```
e.g.
```
pub(crate) const FOREVER: &str = "forever";
```

Then using this const will help find places that reference it later.

cgwalters commented on rust/src/lib/ip.rs:
```diff
@@ -1213,3 +1288,17 @@ fn sanitize_ipv6_token_to_string(
 }
+
+fn is_ip_addrs_none_or_all_auto(addrs: Option<&[InterfaceIpAddr]>) -> bool {
+    if let Some(addrs) = addrs {
```
```suggestion
addrs.map_or(false, |addrs| ...)
```

cgwalters reviewed (commented):
Just gave this a skim, didn't review for any functional changes

cathay4t committed:
ip: Support static IP address on with auto ip stack

Introducing two properties for IP address:
 * `valid-left`: seconds for IP address been valid. `forever` means
   static.
 * `preferred_left`: seconds for IP address been preferred. `forever`
   means static.

With above properties, we can identify which certain IP address is
retrieved from DHCP/Autoconf so we don't convert dynamic IP address to
static when user apply the queried state back.

With this patch, user is allowed to set static IP address with
`dhcp: true` and/or `autoconf: true`.

When desire state has DHCP/Autoconf disabled but all desired IP
addresses are dynamic, nmstate will convert those dynamic IP to static.
This is common use case where user just copy retrieved dynamic IP and
hoping nmstate can convert it to static.

Integration test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>
